### PR TITLE
[16.01] libraries: handle correctly application/json pased in POST

### PR DESCRIPTION
**Pull request by martenson**
This is an attempt to properly fix: https://github.com/galaxyproject/galaxy/pull/1629

ping @gregvonkuster @blankenberg

### CODE REVIEW

**Comment by jmchilton** [comment]
I preferred Greg's documentation changes - it seemed more inline with most of the rest of the API. Still this is a good change on its own.

:+1:

**Comment by martenson** [comment]
You can create a POST request with url parameters, then the payload is non-existent, but the API should still work imo.

**Comment by jmchilton** [comment]
For nearly all other Galaxy POST requests - we consume request properties via payload and ignore URL parameters. If it needs to work for this endpoint for backward compatibility that is totally fine - but we shouldn't encourage it IMO.

**Comment by martenson** [comment]
@gregvonkuster please test this PR with your implementation thanjks

**Comment by gregvonkuster** [comment]
@martenson This fix worked for my post.  You may need to make similar changes in other POST methods for libraries.  Thanks Martin!  ;)

**Comment by martenson** [comment]
updated and rebased

I tried to fix all library API endpoints accepting POST, PUT, PATCH, DELETE where it made sense to accept json

I also fixed a remnant of poor Role handling in folders API controller, it now uses Role manager similar to libraries controller

File: lib/galaxy/webapps/galaxy/api/libraries.py [hunk 1]
```diff
@@ -239,7 +240,7 @@ def get_permissions( self, trans, encoded_library_id, **kwd ):
             raise exceptions.RequestParameterInvalidException( "The value of 'scope' parameter is invalid. Alllowed values: current, available" )
 
     @expose_api
-    def set_permissions( self, trans, encoded_library_id, **kwd ):
+    def set_permissions( self, trans, encoded_library_id, payload=None, **kwd ):
         """
         def set_permissions( self, trans, encoded_dataset_id, **kwd ):
```
Comment (nsoranzo) [inline lib/galaxy/webapps/galaxy/api/libraries.py]: This doc line is not in sync with the method signature, I'd suggest to remove it.

Comment (martenson) [inline lib/galaxy/webapps/galaxy/api/libraries.py]: will fix

**Comment by martenson** [comment]
updated documentation of all touched methods

**Comment by martenson** [comment]
please review and merge, this needs to go to release

**Comment by martenson** [comment]
thanks John

**Pull request merged**